python/ray/data/datasource/file_based_datasource.py

```diff
@@ -354,12 +370,20 @@ def write(
                 for row_index, row in enumerate(
                     block.iter_rows(public_row_format=False)
                 ):
-                    # TODO: Refactor `BlockWritePathProvider` to support rows.
-                    filename = (
-                        f"{dataset_uuid}_{ctx.task_idx:06}_{block_idx:06}_"
-                        f"{row_index:06}.{file_format}"
-                    )
-                    write_path = os.path.join(path, filename)
+                    if filename_provider is not None:
+                        file_index = _generate_unique_file_index(
```

Comment:
Would it make sense to make this function part of the interface as well? I could imagine scenarios where it'd be useful to have the raw index (e.g., to keep files sorted).

Reply:
Actually another thought is to just allow get_filename_for_row to take in this tuple, so users don't have to work around the abstraction if they do need the index. I think it's okay to expose this since we need to expose blocks anyway.